Test/Sampler/test_autoreg.py

```diff
@@ -0,0 +1,37 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import jax
+import netket as nk
+import pytest
+from jax import numpy as jnp
+
+
+@pytest.mark.parametrize("s", [1 / 2, 1])
+def test_ARDirectSampler(s):
+    L = 4
+    n_chains = 3
```

Comment:
can't you add this to the sampling setup in `Test/Sampler/test_sampler.py`, where you only have to declare it at the beginning or that setup won't work with ARNN?

Reply:
I think it's different. `ARDirectSampler` only works with `ARNN` (or something that implements `conditionals`), not RBM.

I copied `test_states_in_hilbert` from `Test/Sampler/test_sampler.py` to `test_autoreg.py`. For `test_correct_sampling`, we need to first train the network, then compare the samples with the true distribution. (Which may be the scope of `Test/Variational/test_autoreg.py`, should I write it there?) For `test_throwing`, those tests are only for `MetropolisHamiltonian*`.

Reply:
in ```test_correct_sampling``` you don't need to train the network, you can just use random weights and check that the sampled distribution agrees with the exactly sampled one

Reply:
indeed. We can change the fixture `rbm_and_weights(request)` to introspect into the calling test function and return an ARNN instead of an RBM if the sampler is an ARSampler.

I know it's possible to do (google pytest fixture introspection). I'm not sure how to do it though.

Reply:
Ok, I can test if the result from ARNN + Metropolis samplers are consistent with exact sampling. But there is no need to test `ARDirectSampler` with a 'more exact' sampler, because `ARDirectSampler` already exactly samples from the distribution represented by the ARNN (whether trained or not).

Reply:
We simply test that the samples are distributed according to psi^2.
It makes sense. Even if ARDirectSampler samples from it, maybe there might be bugs and this checks them.
We do the same for ExactSampler.